### PR TITLE
Add progress entries contract to openapi v2

### PR DESCRIPTION
**Pull request by BarryPiccinni**

### CODE REVIEW

File: openapi/openapi-v2.json [hunk 1]
```diff
@@ -422,6 +425,602 @@
...
+                            "additionalProperties": false,
+                            "properties": {
+                                "position": {
+                                    "enum": ["current", "first"]
+                                },
+                                "sectionId": {
+                                    "title": "Section Id",
+                                    "type": "string",
+                                    "pattern": "^(?:p-?(?:-[a-z0-9]{1,20}){1,20}|system|referrer)$"
+                                }
```
Comment (armoj) [inline openapi/openapi-v2.json:463]: are we removing `referrer` from `sectionId`?

File: openapi/openapi-v2.json [hunk 2]
```diff
@@ -422,6 +425,602 @@
...
+                                                                        "type": "string"
+                                                                    },
+                                                                    "id": {
+                                                                        "type": "string",
+                                                                        "anyOf": [
+                                                                            {
+                                                                                "title": "Section Id",
+                                                                                "type": "string",
+                                                                                "pattern": "^(?:p-?(?:-[a-z0-9]{1,20}){1,20}|system|referrer)$"
+                                                                            },
```
Comment (armoj) [inline openapi/openapi-v2.json:584]: are we getting rid of referrer?

File: openapi/openapi-v2.json [hunk 3]
```diff
@@ -422,6 +425,602 @@
...
+                                                                            "sectionId": {
+                                                                                "anyOf": [
+                                                                                    {
+                                                                                        "nullable": true
+                                                                                    },
+                                                                                    {
+                                                                                        "title": "Section Id",
+                                                                                        "type": "string",
+                                                                                        "pattern": "^(?:p-?(?:-[a-z0-9]{1,20}){1,20}|system|referrer)$"
+                                                                                    }
```
Comment (armoj) [inline openapi/openapi-v2.json:627]: are we getting rid of referrer?

File: questionnaire/routes.test.js [hunk 4]
```diff
@@ -434,4 +435,110 @@ describe('V2 openapi validation', () => {
...
+        const token =
+            'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJhdWQiOiJkYXRhLWNhcHR1cmUtc2VydmljZSIsImlzcyI6IiQuYXVkIiwianRpIjoiYWE3Nzk1ZmItNDg2Yy00NWEwLWJkNGMtZTMwNjFlNmNjNDk2Iiwic3ViIjoiY2ljYS13ZWIiLCJzY29wZSI6ImNyZWF0ZTpxdWVzdGlvbm5haXJlcyByZWFkOnF1ZXN0aW9ubmFpcmVzIHVwZGF0ZTpxdWVzdGlvbm5haXJlcyBkZWxldGU6cXVlc3Rpb25uYWlyZXMgcmVhZDpwcm9ncmVzcy1lbnRyaWVzIHJlYWQ6YW5zd2VycyIsImlhdCI6MTY4MDcwNTI3N30.OFXEk5CjaMZJVmS8Ioke2l2AlffayMCvIWZ2DwJCu2o';
+
+        it('should return status code 401 if bearer token is NOT valid', async () => {
+            const response = await request(app)
+                .get('/api/questionnaires/285cb104-0c15-4a9c-9840-cb1007f098fb/progress-entries')
+                .set('Authorization', `Bearer I-AM-INVALID`)
+                .set('Content-Type', 'application/vnd.api+json')
+                .set('Dcs-Api-Version', '2023-05-17')
+                .set('on-behalf-of', `urn:uuid:f81d4fae-7dec-11d0-a765-00a0c91e6bf6`);
```
Comment (armoj) [inline questionnaire/routes.test.js:449]: title case the header name here

File: questionnaire/routes.test.js [hunk 5]
```diff
@@ -434,4 +435,110 @@ describe('V2 openapi validation', () => {
...
+        it('should return status code 403 if the bearer token has insufficient scope', async () => {
+            // THIS IS A TOKEN WITH A DUMMY SCOPE
+            const dummyToken =
+                'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJhdWQiOiJkYXRhLWNhcHR1cmUtc2VydmljZSIsImlzcyI6IiQuYXVkIiwianRpIjoiNTFhODljYWUtM2Q1MC00ZDc1LTliMmEtMjU2NzliODgwMTkxIiwic3ViIjoiY2ljYS13ZWIiLCJzY29wZSI6ImNyZWF0ZTpub3RoaW5nIiwiaWF0IjoxNjgwNzk4NDU5fQ.97LgtlW_dcAV0Xno6BsbVmuyhLtq4gCoVWGQ56_VmEk';
+            const response = await request(app)
+                .get('/api/questionnaires/285cb104-0c15-4a9c-9840-cb1007f098fb/progress-entries')
+                .set('Authorization', `Bearer ${dummyToken}`)
+                .set('Content-Type', 'application/vnd.api+json')
+                .set('Dcs-Api-Version', '2023-05-17')
+                .set('on-behalf-of', `urn:uuid:f81d4fae-7dec-11d0-a765-00a0c91e6bf6`);
```
Comment (armoj) [inline questionnaire/routes.test.js:464]: title case the header name here

File: questionnaire/routes.test.js [hunk 6]
```diff
@@ -434,4 +435,110 @@ describe('V2 openapi validation', () => {
...
+
+        it('should return status code 404 if the query string contains incorrect data', async () => {
+            const response = await request(app)
+                .get(
+                    '/api/questionnaires/285cb104-0c15-4a9c-9840-cb1007f098fb/progress-entries?filter[sectionId]=p--not-a-valid-section'
+                )
+                .set('Authorization', `Bearer ${token}`)
+                .set('Content-Type', 'application/vnd.api+json')
+                .set('Dcs-Api-Version', '2023-05-17')
+                .set('on-behalf-of', `urn:uuid:f81d4fae-7dec-11d0-a765-00a0c91e6bf6`);
```
Comment (armoj) [inline questionnaire/routes.test.js:478]: title case the header name here

File: questionnaire/routes.test.js [hunk 7]
```diff
@@ -434,4 +435,110 @@ describe('V2 openapi validation', () => {
...
+                    .get(
+                        '/api/questionnaires/285cb104-0c15-4a9c-9840-cb1007f098fb/progress-entries'
+                    )
+                    .set('Authorization', `Bearer ${token}`)
+                    .set('Content-Type', 'application/vnd.api+json')
+                    .set('Dcs-Api-Version', '2023-05-17');
+                expect(response.body).toHaveProperty('errors');
+                expect(response.body.errors[0].status).toEqual(400);
+                expect(response.body.errors[0].detail).toEqual(
+                    "should have required property 'on-behalf-of'"
```
Comment (armoj) [inline questionnaire/routes.test.js:498]: title case the header name here

File: questionnaire/routes.test.js [hunk 8]
```diff
@@ -434,4 +435,110 @@ describe('V2 openapi validation', () => {
...
+                );
+            });
+
+            it('should return status code 200 if owner data is included in the header', async () => {
+                const response = await request(app)
+                    .get(
+                        '/api/questionnaires/285cb104-0c15-4a9c-9840-cb1007f098fb/progress-entries'
+                    )
+                    .set('Authorization', `Bearer ${token}`)
+                    .set('on-behalf-of', `urn:uuid:f81d4fae-7dec-11d0-a765-00a0c91e6bf6`)
```
Comment (armoj) [inline questionnaire/routes.test.js:508]: title case the header name here

File: openapi/src/json-schemas/models/definitions/owner-id.json [hunk 9]
```diff
@@ -1,6 +1,6 @@
 {
     "$schema": "http://json-schema.org/draft-07/schema#",
-    "title": "Owner Id",
+    "title": "On-behalf-of",
```
Comment (armoj) [inline openapi/src/json-schemas/models/definitions/owner-id.json:3]: What's the reason behind this title change? Is it to make it clear that this owner ID directly maps to the on-behalf-of header? Is it not more correct to be called `Owner Id` as it already is? Just a thought

**Review by armoj — CHANGES_REQUESTED** [review]

File: questionnaire/routes.test.js (hunk 7 above)
Comment (BarryPiccinni) [reply questionnaire/routes.test.js:498]: The error transforms the header to lower-case from some reason so this is correct.

File: openapi/src/json-schemas/models/definitions/owner-id.json (hunk 9 above)
Comment (BarryPiccinni) [reply openapi/src/json-schemas/models/definitions/owner-id.json]: Yep, I was confusing this with the header title. Have changed it back.

File: openapi/openapi-v2.json (hunk 1 above)
Comment (armoj) [reply openapi/openapi-v2.json:463]: Is there a test to check that system/referrer/owner are valid, and any other word is not valid

**Review by armoj — CHANGES_REQUESTED** [review]

File: openapi/openapi-v2.json (hunk 1 above)
Comment (BarryPiccinni) [reply openapi/openapi-v2.json:463]: We've got a test that says it should 404 if it doesn't match that regex, but not specifically testing those 3 values

**Review by armoj — APPROVED** [review]

**Pull request merged**